implementations/rust/ockam/ockam_vault_aws/src/vault/aws_security_module.rs

```diff
@@ -52,13 +100,105 @@ impl SecurityModule for AwsSecurityModule {
     async fn verify(
         &self,
         public_key: &PublicKey,
         message: &[u8],
         signature: &Signature,
     ) -> Result<bool> {
-        let key_id = self.get_key_id(public_key).await?;
-        self.client.verify(&key_id, message, signature).await
+        use p256::ecdsa::{signature::Verifier as _, Signature, VerifyingKey};
+
+        let verifying_key =
+            VerifyingKey::from_public_key_der(public_key.data()).map_err(Self::from_pkcs8)?;
+        let ecdsa_signature = Signature::from_der(signature.as_ref()).map_err(Self::from_ecdsa)?;
+        Ok(verifying_key.verify(message, &ecdsa_signature).is_ok())
+    }
+}
+
+impl AwsSecurityModule {
+    pub(crate) fn from_ecdsa(e: p256::ecdsa::Error) -> Error {
+        Error::new(Origin::Vault, Kind::Unknown, e)
+    }
+
+    pub(crate) fn from_pkcs8<T: core::fmt::Display>(e: T) -> Error {
+        #[cfg(feature = "no_std")]
+        use ockam_core::compat::string::ToString;
+        Error::new(Origin::Vault, Kind::Unknown, e.to_string())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ockam_node::InMemoryKeyValueStorage;
+
+    /// This test needs to be executed with the following environment variables
+    /// AWS_REGION
+    /// AWS_ACCESS_KEY_ID
+    /// AWS_SECRET_ACCESS_KEY
+    #[tokio::test]
+    #[ignore]
+    async fn test_store_public_key_key_id_mapping() -> Result<()> {
+        let storage = InMemoryKeyValueStorage::create();
+        let security_module =
+            AwsSecurityModule::new(AwsKmsConfig::default().await?, storage.clone()).await?;
+
+        let key_id = security_module
+            .create_secret(SecretAttributes::NistP256)
+            .await?;
+
+        // the public key can be retrieved using the kms client directly
+        // but then the public key <-> key id mapping is not cached
+        let public_key = security_module.client.public_key(&key_id).await;
+        assert!(public_key.is_ok());
+        assert!(storage.get(&public_key?).await?.is_none());
+
+        // when the public key is retrieved using the security module
+        // then the public key <-> key id mapping is cached locally
+        let public_key = security_module.get_public_key(&key_id).await;
+        assert!(public_key.is_ok());
+
+        let public_key = public_key?;
+        let key_id = storage.get(&public_key).await;
+        assert!(key_id.is_ok());
+
+        let key_id = security_module.get_key_id(&public_key).await;
+        assert!(key_id.is_ok());
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn test_sign_verify() -> Result<()> {
+        let security_module = AwsSecurityModule::default().await?;
+        let key_id = security_module
+            .create_secret(SecretAttributes::NistP256)
+            .await?;
+        let message = b"hello world";
+        let signature = security_module.sign(&key_id, &message[..]).await?;
+
+        // Verify locally
```

Comment:
Aren't those 2 calls identical?

Reply:
That's the thing tested here:

 - the first one uses a local implementation `security_module.sign`
 - the second one uses a remote implementation via the client `security_module.client.sign`

This test is a sanity check that both methods return the same result.

Reply:
So why do both verify calls use `security_module.client`?